modules/automode.cpp

```diff
@@ -0,0 +1,693 @@
+    }
+    void OnClearCommand() {
+        m_msUsers.clear();
+        PutModule(t_s("All entries cleared."));
+        ClearNV(true);
+        SaveRegistry();
+    }
+    ~CAutoModeMod() override {
+        for (const auto& it : m_msUsers) {
+            delete it.second;
+        }
+        m_msUsers.clear();
+    }
+
+    void OnJoin(const CNick& Nick, CChan& Channel) override {
+        // If we have ops in this chan
+        if (Channel.HasPerm(CChan::Op)) {
+            CheckAutoMode(Nick, Channel);
+        }
+    }
+
+    void OnQuit(const CNick& Nick, const CString& sMessage,
+                const vector<CChan*>& vChans) override {
+        MCString::iterator it = m_msQueue.find(Nick.GetNick().AsLower());
+
+        if (it != m_msQueue.end()) {
+            m_msQueue.erase(it);
+        }
+    }
+
+    void OnNick(const CNick& OldNick, const CString& sNewNick,
+                const vector<CChan*>& vChans) override {
+        // Update the queue with nick changes
+        MCString::iterator it = m_msQueue.find(OldNick.GetNick().AsLower());
+
+        if (it != m_msQueue.end()) {
+            m_msQueue[sNewNick.AsLower()] = it->second;
+            m_msQueue.erase(it);
+        }
+    }
+
+    EModRet OnPrivNotice(CNick& Nick, CString& sMessage) override {
+        if (!sMessage.Token(0).Equals("!ZNCAO")) {
+            return CONTINUE;
+        }
+
+        CString sCommand = sMessage.Token(1);
+
+        if (sCommand.Equals("CHALLENGE")) {
+            ChallengeRespond(Nick, sMessage.Token(2));
+        } else if (sCommand.Equals("RESPONSE")) {
+            VerifyResponse(Nick, sMessage.Token(2));
+        }
+
+        return HALTCORE;
+    }
+
+    void OnOp2(const CNick* pOpNick, const CNick& Nick, CChan& Channel,
+               bool bNoChange) override {
+        if (Nick.GetNick() == GetNetwork()->GetIRCNick().GetNick()) {
+            const map<CString, CNick>& msNicks = Channel.GetNicks();
+
+            for (const auto& it : msNicks) {
+                if (!it.second.HasPerm(CChan::Op)) {
+                    CheckAutoMode(it.second, Channel);
+                }
+            }
+        }
+    }
+
+    void OnModCommand(const CString& sLine) override {
+        CString sCommand = sLine.Token(0).AsUpper();
+        if (sCommand.Equals("TIMERS")) {
+            // for testing purposes - hidden from help
+            ListTimers();
+        } else {
+            HandleCommand(sLine);
+        }
+    }
+
+    void OnAddUserCommand(const CString& sLine) {
+        CString sUser = sLine.Token(1);
+        CString sHost = sLine.Token(2);
+        CString sMode = sLine.Token(3);
+        CString sKey = sLine.Token(4);
+
+        if (sHost.empty()) {
+            PutModule(
+                t_s("Usage: AddUser <user> <hostmask>[,<hostmasks>...] <mode> <key> "
+                    "[channels]"));
+        } else if (sMode != "q" && sMode != "a" && sMode != "o" &&
+                   sMode != "h" && sMode != "v") {
```

Comment:
I suggest reading it from PREFIX ISUPPORT instead. Or even skip this check and leave it to user, because ISUPPORT can change if e.g. user connects to a different server